### PR TITLE
[TRTLLM-13212][refactor] Unify sampler ops and clean up dead code in TorchSampler

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Summary by CodeRabbit

* **New Features**
  * Unified sampling APIs now support top-k, top-p, temperature, generators, and seed/offset-based randomness.
  * Added sampling options for top-k-only and top-p-only selection from probabilities.
  * Improved CUDA graph compatibility with stateless random seeds and offsets.

* **Bug Fixes**
  * Improved output allocation metadata handling.
  * Corrected edge cases for sampling temperatures, log probabilities, finish reasons, and minimum-length penalties.
  * Updated speculative decoding sampling behavior for consistent top-k/top-p handling.

* **Refactor**
  * Removed the standalone probability-computation operation in favor of the unified sampling flow.

## Description

Follow-ups promised in the #15542 review (sampler ops centralization), plus dead-code
removal and small correctness/readability fixes found while auditing the sampler module.
All changes are behavior-preserving unless noted.

### Dedup thin wrappers (ops/vanilla.py) — #15542 review follow-up

- Fold the `top_k_/top_p_/temperature_sampling_batch` pass-through wrappers into
  `top_k_top_p_sampling_batch` via default arguments (`top_k=None` disables top-k,
  `top_p=1` disables top-p).
- Drop the pass-through `greedy()` wrapper; the `greedy` re-export now aliases
  `greedy_search_sampling_batch` directly.

### Merge flashinfer op variants (ops/flashinfer.py) — #15542 review follow-up

- Merge the seed/offset and generator variants of each sampling op wrapper into a
  single signature with keyword-only randomness arguments (6 wrappers -> 4).
  flashinfer's own API (>= 0.6.4) already accepts both in one signature, with explicit
  `seed`/`offset` taking precedence over `generator`; the split wrappers predated that.
- Both randomness modes are preserved: `seed`/`offset` (device tensors) remain required
  for CUDA-graph capture and cross-rank determinism; `generator` remains for eager paths.
  The module docstring now documents this distinction.

### Remove dead code

- `trtllm::compute_probs_from_logits_op` chain, unused end to end (no Python or C++
  callers): the `register_fake` stub, the thop wrapper + `TORCH_LIBRARY` registration,
  and the exclusive kernel chain in `dynamicTreeKernels.cu` (`topKProbStage1/2`,
  `invokeTopKTopPMaskingForProbs`, `applyTopKTopPForProbOp`, `computeSoftmaxForProbOp`,
  `computeProbsFromLogits`) plus the torch/ATen includes only they used.
- Four `TorchSampler` methods with zero callers: `_handle_finish_reasons`,
  `_longest_stop_word_len`, `_requests_with_stop_words`, `_request_indices_with_stop_words`.

### Promote de-facto public internals

`sampling_utils` and `sampler.py` were reaching into underscore-private members of
`ops.vanilla`; promote the actually-shared surface and remove a duplicated constant:

- `_GREEDY_TEMPERATURE_THRESHOLD` -> `GREEDY_TEMPERATURE_THRESHOLD`; `DISABLE_TEMP_VAL`
  in `speculative/interface.py` now derives from it (`/ 10`, value unchanged at 1e-5)
  so the sentinel and threshold cannot drift apart silently.
- `_safely_apply_temperature_inplace` -> `safely_apply_temperature_inplace`,
  `_Fusions` -> `Fusions`.
- Single definition for `BEAM_SEARCH_PAD_TOKEN` (in `ops.vanilla`, re-exported by the
  facade); previously defined independently in two files.
- `sampling_utils` is now the only module importing the `ops.*` backend modules.

### Correctness / readability fixes in TorchSampler

- `_get_logprobs_from_request`: use forward slicing for the history view;
  `[:, :-preallocate_extra_steps, :]` silently yields an **empty** view when
  `preallocate_extra_steps == 0` (the default), skipping history filling.
- `_apply_min_length_penalty`: return `None` instead of echoing the in-place-mutated
  logits; defer the host tensor->list conversions until after the early-out (the common
  no-min-length path now skips them); flatten double-nested guard `if`s into early
  `continue`s, hoisting the per-request offset accumulation above the guards.
- `_flashinfer_check_nans`: restore the explanatory comments lost in the #15542 move —
  the constant `False` deliberately keeps flashinfer's *syncing* `check_nan` path
  disabled while the async device-side assert provides NaN protection without stalling
  the pipeline (flashinfer issue #1575).
- Fix `seed`/`offset` type hints on `sampling_batch_spec_dec_one_model`
  (`Optional[int]` -> `Optional[torch.Tensor]`; callers pass device tensors).

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by QiJune — APPROVED** [review]
LGTM

**Review by QiJune — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/flashinfer.py [hunk 1]
```diff
@@ -27,119 +36,146 @@
...
         deterministic=True,
         check_nan=check_nan,
         generator=generator,
+        seed=seed,
+        offset=offset,
     )
     return tokens
 
 
+# The three ops below wrap the mask -> softmax -> renorm pipeline stages 1:1.
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/ops/flashinfer.py:147]: For instance, `softmax_op` will fail if FlashInfer is not available. Importing something that certainly fails if it ever gets invoked may result in late failures (imagine a service being "healthy" and crashing only after a user request selects some unsupported behavior). It might be preferable if that service failed at startup time already.

File: tensorrt_llm/_torch/pyexecutor/sampler/sampling_utils.py [hunk 2]
```diff
@@ -31,52 +31,45 @@
 # These op wrappers are safe to import without flashinfer installed; they are
 # only called on the flashinfer sampler / speculative-worker paths.
 from tensorrt_llm._torch.pyexecutor.sampler.ops.flashinfer import (
-    sampling_from_probs_generator_op as sampling_from_probs_generator_op,
+    sampling_from_probs_op as sampling_from_probs_op,
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampling_utils.py:34]: Replace `from a import b as b` with `from a import b` (multiple occurrences).

File: tensorrt_llm/_torch/pyexecutor/sampler/sampling_utils.py [hunk 3]
```diff
@@ -814,7 +825,7 @@ def sample_grouped_strategies(
 
 
 # Re-export the torch greedy op (used by drafting_loops and speculative/interface).
-greedy = _torch_greedy
+greedy = greedy_search_sampling_batch
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampling_utils.py:828]: `greedy_search_sampling_batch` is a better name. Should we perhaps change the dependent modules to use this name instead?

File: tensorrt_llm/_torch/pyexecutor/sampler/sampling_utils.py [hunk 4]
```diff
@@ -825,12 +836,11 @@ def sanitize_top_k(top_k: torch.Tensor, vocab_size: int) -> torch.Tensor:
...
-    non-positive value (and any oversized disable sentinel such as
-    ``INT32_MAX``) to ``vocab_size`` (== keep all tokens), leaving genuine
-    top_k values untouched.
+    but the flashinfer top-k kernels (``top_k_mask_logits``) break on a literal
+    0 — they mask the entire row (all-zero probs). Map any non-positive value
+    (and any oversized disable sentinel such as ``INT32_MAX``) to
+    ``vocab_size`` (== keep all tokens), leaving genuine top_k values
+    untouched.
     """
     return torch.where(top_k > 0, top_k, torch.full_like(top_k, vocab_size)).clamp(max=vocab_size)
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampling_utils.py:845]: Consider using `.masked_fill_`.

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/flashinfer.py (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/ops/flashinfer.py:147]: Thanks for the review, @ixlmar. Since this PR has already been merged, I'll address your suggestions in [#16052](https://github.com/NVIDIA/TensorRT-LLM/pull/16052)
